CMakeLists.txt

```diff
@@ -727,17 +727,33 @@ add_library(hack_bytecode_handlers_ir EXCLUDE_FROM_ALL util/execution/bytecode_h
 target_link_libraries(hack_bytecode_handlers_ir PRIVATE noisepage_static)
 set_target_properties(hack_bytecode_handlers_ir PROPERTIES CXX_EXTENSIONS OFF)
 
+# On MacOS, the clang++ we want is not the clang++ in PATH. On Linux, it is.
+# NOTE(Kyle): It is important that CLANG is resolved prior to the commands
+# below, so the location of the following definition matters.
+if (APPLE)
+    set(CLANG "${LLVM_TOOLS_BINARY_DIR}/clang++")
+else ()
+    find_program(CLANG NAMES "clang++-8" "clang++" REQUIRED)
+endif ()
+
+# Run the build-support script that generates and optimizes bitcode.
+add_custom_command(
+        OUTPUT "${CMAKE_BINARY_DIR}/bin/bytecode_handlers_ir.bc"
+        DEPENDS gen_opt_bc "${PROJECT_SOURCE_DIR}/src/include/execution/vm/bytecodes.h" "${PROJECT_SOURCE_DIR}/src/include/execution/vm/bytecode_handlers.h"
+        WORKING_DIRECTORY ${CMAKE_BINARY_DIR}
+        COMMAND ${BUILD_SUPPORT_DIR}/generate-bitcodes.sh ${PROJECT_SOURCE_DIR} ${CMAKE_BINARY_DIR} ${CLANG}
+)
+
+# Here, we create a custom target that merely wraps the output file
+# from the custom command above; this ensures that the custom command
+# is actually only ever invoked once even when multiple indepdendent
+# targets that require bitcode are built because the build system 
+# (e.g. Make, Ninja) is now 'aware' of the target-level dependency.
+add_custom_target(bitcode_generator DEPENDS "${CMAKE_BINARY_DIR}/bin/bytecode_handlers_ir.bc")
+
+# A declarative convenience wrapper for expressing dependency on bitcode.
 function(target_requires_bitcode TARGET)
```

Comment:
Since this function doesn't do anything else, does it make sense to just directly write `add_dependencies(... bitcode_generator)` wherever necessary? Seems clearer.

Reply:
Only argument I have for maybe keeping it around was that it made this refactor much easier - all I had to do was hack around on stuff that would then populate the body of this function, leaving all of the "call sites" unmodified. This has the potential to make future updates less invasive.

But I see your point, because in it's current state it looks like this might be hiding some complexity, when in reality it actually does just the one thing. We might need a tiebreaker vote.

Reply:
I guess the question is, do you foresee future updates? If yes, happy to merge as is, otherwise I feel anticipating use cases has generally gone poorly / added indirection where it didn't help much.

Reply:
All code would be ctrl-f'able on `bitcode_generator`.

Reply:
Pushed the refactor.